src/ucp/api/ucp_def.h

```diff
@@ -81,6 +82,30 @@ typedef struct ucp_config                ucp_config_t;
 typedef struct ucp_ep                    *ucp_ep_h;
 
 
+/**
+ * @ingroup UCP_ENDPOINT
+ * @brief Callback to process incoming active message
+ *
+ * @param [in]  arg     User-defined argument.
+ * @param [in]  data    Points to the received data. This may be part of
```

Comment:
- currently UCP doesn't have transport level limitations such as max message length. it means that UCP should allocate memory and do extra copying for big messages => performance won't be optimal, right?
 - is some API missed for releasing of the descriptor?

Reply:
There is no API missing for the releasing of the descriptor. I need to change the description of the data parameter, which I will do now. As specified in the @retval section of description, AM's at the UCP level must return UCS_OK. This ensures descriptors are never created and, therefore, never need to be released. For example, 
https://github.com/openucx/ucx/blob/master/src/uct/ib/base/ib_iface.h#L206

We did not implement zcopy_multi or bcopy_multi, so a user is limited by max message length. If a user does pass in a length that triggers either a zcopy_multi or bcopy_multi, an error is shown to the user and UCS_ERR_NOT_SUPPORTED is returned. 

We thought that generally active messages would be used for smaller messages that would not test the max message length, although we are aware of the issue. We also thought that the sending of one active messages invoking several active messages on the remote endpoint would be difficult for the user to handle, which would happen with zcopy_multi or bcopy_multi. However, we are open to revisiting this if it is deemed necessary.

Reply:
If bcopy_multi or zcopy_multi is called, it will return UCS_ERR_NOT_IMPLEMENTED. It would not return UCS_ERR_NOT_SUPPORTED as my last comment stated.

Reply:
"We also thought that the sending of one active messages invoking several active messages on the remote endpoint would be difficult for the user to handle" - it does not have to be this way. You can invoke the callback once full message is shipped (and yes it will require additional ucp logic)

Reply:
Just to clarify, you are taking about invoking the callback only once all of the data is received? You are not referring to the invocation of the completion callback when the data has been sent. I am confused by "shipped".